static/app/utils/profiling/profile/eventedProfile.tsx

```diff
@@ -0,0 +1,165 @@
+
+      if (!frame) {
+        throw new Error(`Cannot retrieve event: ${event.frame} from frame index`);
+      }
+
+      switch (event.type) {
+        case 'O': {
+          profile.enterFrame(frame, event.at);
+          break;
+        }
+        case 'C': {
+          profile.leaveFrame(frame, event.at);
+          break;
+        }
+        default: {
+          throw new TypeError(`Unknown event type ${event.type}`);
+        }
+      }
+    }
+
+    return profile.build();
+  }
+
+  addWeightToFrames(weight: number): void {
+    const weightDelta = weight - this.lastValue;
+
+    for (const frame of this.stack) {
+      frame.addToTotalWeight(weightDelta);
+    }
+
+    const top = lastOfArray(this.stack);
+    if (top) top.addToSelfWeight(weight);
+  }
+
+  addWeightsToNodes(value: number) {
+    const delta = value - this.lastValue;
+
+    for (const node of this.appendOrderStack) {
+      node.addToTotalWeight(delta);
+    }
+    const stackTop = lastOfArray(this.appendOrderStack);
+
+    if (stackTop) {
+      stackTop.addToSelfWeight(delta);
+    }
+  }
+
+  enterFrame(frame: Frame, at: number): void {
+    this.addWeightToFrames(at);
+    this.addWeightsToNodes(at);
+
+    const lastTop = lastOfArray(this.appendOrderStack);
+
+    if (lastTop) {
+      const sampleDelta = at - this.lastValue;
+
+      // If the sample timestamp is not the same as the same as of previous frame,
+      // we can deduce that this is a new sample and need to push it on the stack
+      if (sampleDelta > 0) {
+        this.samples.push(lastTop);
+        this.weights.push(sampleDelta);
+      }
+
+      const last = lastOfArray(lastTop.children);
+      let node: CallTreeNode;
+
+      if (last && !last.isLocked() && last.frame === frame) {
+        node = last;
+      } else {
+        node = new CallTreeNode(frame, lastTop);
+        lastTop.children.push(node);
+      }
+
+      let start = this.appendOrderStack.length - 1;
+
+      // This is On^2, because we iterate over all frames in the stack to check if our
+      // frame is a recursive frame. We could do this in O(1) by keeping a map of frames in the stack
+      // We check the stack in a top-down order to find the first recursive frame.
+      while (start >= 0) {
+        if (this.appendOrderStack[start]?.frame === node.frame) {
+          node.setRecursive(this.appendOrderStack[start]);
+          break;
+        }
+        start--;
+      }
+
+      this.appendOrderStack.push(node);
+    }
+
+    this.stack.push(frame);
+    this.lastValue = at;
+  }
+
+  leaveFrame(_event: Frame, at: number): void {
+    this.addWeightToFrames(at);
+    this.addWeightsToNodes(at);
+
+    const leavingStackTop = this.appendOrderStack.pop();
+    if (!leavingStackTop) throw new Error('Unbalanced stack');
+
+    // Lock the stack node, so we make sure we dont mutate it in the future.
+    // The samples should be ordered by timestamp when processed so we should never
```

Comment:
Should we throw an error if they are not ordered by timestamp?

Reply:
@AbhiPrasad Yeah, I dont think I ever saw this happen, but we should at least know if it does!